dev/com.ibm.ws.security.kerberos.auth/src/com/ibm/ws/security/kerberos/auth/internal/LRUCache.java

```diff
@@ -84,37 +86,38 @@ public Subject get(KerberosPrincipal principal) {
-                        long renewTillMs = kerbTicket.getRenewTill().getTime();
-                        long timeLeft = renewTillMs - System.currentTimeMillis() - TimeUnit.MINUTES.toMillis(10);
-                        if (timeLeft < 0) {
+                if (kerbTicket.isCurrent()) {
+                    // we have a valid ticket
+                    break;
+                }
+
+                if (kerbTicket.isRenewable()) {
+                    // If the ticket is not current but is renewable, try to renew the ticket if
+                    // there is at least 10m remaining on the renewTill time
+                    Instant renewTill = kerbTicket.getRenewTill().toInstant().minus(10, ChronoUnit.MINUTES);
+                    if (Instant.now().isAfter(renewTill)) {
+                        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+                            Tr.debug(tc, "no match because is renewable but past the renewTill time. renewTill=" + renewTill);
+                        return null;
+                    } else {
+                        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+                            Tr.debug(tc, "found match for non-current ticket, but ticket is renewable. Attempting to renew now");
+                        try {
+                            kerbTicket.refresh();
+                        } catch (RefreshFailedException e) {
                             if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
-                                Tr.debug(tc, "no match because is renewable but past the renewTill time. timeLeft=" + timeLeft);
+                                Tr.debug(tc, "unable to refresh kerberos ticket due to: " + e);
                             return null;
-                        } else {
-                            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
-                                Tr.debug(tc, "found match for non-current ticket, but ticket is renewable. Attempting to renew now");
-                            oldTicket = kerbTicket;
-                            renewedTicket = cloneTicket(kerbTicket);
-                            break;
                         }
-                    } else {
-                        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
-                            Tr.debug(tc, "no match because ticket is not current and not renewable");
-                        return null;
+                        break;
                     }
+                } else {
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled())
+                        Tr.debug(tc, "no match because ticket is not current and not renewable");
+                    return null;
                 }
             }
-            if (oldTicket != null && renewedTicket != null) {
-                kerbTickets.remove(oldTicket);
```

Comment:
This remove/add code is what needed to be replaced, since `sub.getPrivateCredentials(KerberosTicket.class)` only returns a copy of the credential set, not the actual set of credentials

Reply:
ok, so we need to clone the Kerberos ticket and refresh/renew as needed.

Reply:
If we cloned the ticket and refreshed the cloned ticket then we would need to mutate the set of private credentials on the subject, which I hesitate to do because that could impact existing resources that are using this Subject.

Instead, I think we should just call `kerbTicket.refresh()` and not clone/replace it. I verified in the JDK implementation of `KerberosTicket.refresh()` that the ticket does properly synchronize on itself and clone and re-init its internal state during the refresh operation.